src/operators/aggregate/hash_aggregate.cc

```diff
@@ -569,105 +569,104 @@ void HashAggregate::Finish(Task *ctx) {
   for (int i = 0; i < group_by_cols_.size(); i++) {
     auto group_by = group_by_cols_.at(i);
     auto group_by_type = group_by.type();
-    auto group_by_type_id = group_by_type->id();
 
-    // The following step assume the iteration of tuple map
-    //    is not going to mutate.
-    // TODO: This is a very very tedious construction of the group_by builder.
-    //  Try to optimize it by using the good features within the arrow.
-    switch (group_by_type_id) {
-      case arrow::Type::STRING: {
-        arrow::StringBuilder builder;
-        for (auto const it : *global_tuple_map) {
-          int chunk_id, item_index;
-          auto hash_key = it.first;
-          std::tie(chunk_id, item_index) = it.second;
-          auto raw_chunk = group_by.chunks().at(chunk_id);
-          auto chunk = std::static_pointer_cast<arrow::StringArray>(raw_chunk);
-          // TODO: Optimize this to make it std::move()?
-          std::string value = chunk->GetString(item_index);
-          // TODO: Try to use the UnsafeAppend here,
-          //  but encounter an error.
-          //            builder.UnsafeAppend(value);
-          builder.Append(value);
-        }
-        std::shared_ptr<arrow::Array> array;
-        arrow::Status st = builder.Finish(&array);
-        if (!st.ok()) {
-          std::cerr << "Building the array for " << i
-                    << "th groupby column failed." << std::endl;
-          exit(1);
-        }
-        group_by_arrays.push_back(array);
-        break;
+    // TODO: The following step assume the iteration of tuple map
+    //        is not going to mutate.
+
+    auto data_type = group_by_type;
+
+    // TODO: Possibly abstract this?
+    // Accessor for different types of chunks
+    auto normal_value_accessor = [](const auto &chunk, const auto item_index) {
+      return chunk->Value(item_index);
+    };
+    auto string_value_accessor = [](const auto &chunk, const auto item_index) {
+      return chunk->GetString(item_index);
+    };
+
+    auto construct_group_by_cols = [&]<typename T>(T *, auto value_accessor) {
+      using ArrayType = ArrowGetArrayType<T>;
+      using BuilderType = ArrowGetBuilderType<T>;
+
+      auto builder =
+          std::reinterpret_pointer_cast<BuilderType>(getBuilder(data_type));
+
+      for (auto const it : *global_tuple_map) {
+        int chunk_id, item_index;
+        auto hash_key = it.first;
+        std::tie(chunk_id, item_index) = it.second;
+        auto raw_chunk = group_by.chunks().at(chunk_id);
```

Comment:
I feel we can avoid auto as much as possible for better readability [Optional]

Reply:
That is a good point -- excessive use of auto makes code hard to write. Adding comments to clarify the type can help, but if we don't need auto, we should avoid it.

Reply:
Put this into the todo list in #58, when we refactor hash aggregate.